### ISSUE TITLE
If use with vault, how to delete payment method in drop-in UI?

### ISSUE THREAD

**Issue by coolbeet** [braintree/braintree-ios-drop-in#57]
Currently I dont see a way to let user delete payment method by themselves, is this feature on the product road map?

Thanks!

**Comment by demerino** [comment]
This feature is not on our roadmap at this time. You'll need to delete the payment method on the server side: https://developers.braintreepayments.com/reference/request/payment-method/delete/ruby

**Comment by coolbeet** [comment]
So drop-in UI is just for adding payment method right? 

Technically, one app probably should only have one payment method page where you can both add and delete payment method, this puts drop-in UI in an awkward position that, you have to build your own fully functional payment management page, and after that, there is no need to use drop-in UI since its functionality are all included in the payment management page.

A lot of our users already complain about they could find any place can delete payment method, we have to delete it in the gateway or by api. 

I strongly recommend build payment method deletion functionality in the drop-in UI, probably just let user long press the payment method and delete, just like how you delete app in springboard.

Thanks!

**Comment by demerino** [comment]
I agree that it is awkward and we'll discuss some possible solutions to allow for editing.

One idea we might investigate is for an editing screen that will return a flag along with the nonce that says the users WANTS to delete the payment option. But it would still be up to the developer to implement the removal.

We appreciate the feedback.

**Comment by coolbeet** [comment]
@demerino Thanks!

**Comment by carterbryden** [comment]
Any updates on this?

**Comment by demerino** [comment]
I recently took a pass at this. Let me know what you think of this approach: https://github.com/braintree/braintree-ios-drop-in/pull/89

**Comment by coolbeet** [comment]
@demerino Great job, I like it!

**Comment by demerino** [comment]
**Update:** We're actively working on this feature.

**Comment by Unholny** [comment]
That's great news. Any estimation on when it's going live?

**Comment by demerino** [comment]
I'm hesitant to give an estimate - but we're aiming for sometime in mid June 🤞

**Comment by demerino** [comment]
6.3.0 adds this functionality, set the new flag on the BTDropInRequest to enable

```
request.vaultManager = true
```

**Comment by coolbeet** [comment]
@demerino Awesome!

**Comment by demerino** [comment]
I'll leave this issue open for questions. Let us know what you think. Web and Android will be coming in the next weeks.

**Comment by Nonepse** [comment]
Thanks for this update @demerino! I updated the Pod in my project, but when trying to delete a card, I get an infinite spinner, and the card is not actually deleted:

<img src="https://user-images.githubusercontent.com/5849587/42759534-2b05446e-88ff-11e8-9f85-409d823e6a89.gif" width="200">

- My first suspect was in `BTVaultManagementViewController.m` at [`l.161`](https://github.com/braintree/braintree-ios-drop-in/pull/89/files#diff-c63b5728dec5cc2baa929d240ed4d520R161). In the `UITableView` delegate method called on the cell deletion, a `POST` request is triggered with parameters to indicate which card (identified by its `paymentMethod.nonce`) has to be deleted. The first parameter given to `POST:parameters:completion:` is an empty string `@""`. But if I understand correctly, it turns out that for the `httpType` `BTAPIClientHTTPTypeGraphQLAPI`, the `BTGraphQLHTTP` object is used. It is a subclass of `BTHTTP` and it overrides the `POST` method, marking the endpoint parameter as unused, because GraphQL.

- Breakpointing my way through, I then realized this `POST` request first triggers a `GET` to fetch the merchant configuration, `BTAPIClient` [`l.358`](https://github.com/braintree/braintree_ios/blob/master/BraintreeCore/BTAPIClient.m#L358):
    ```objective-c
    [self fetchOrReturnRemoteConfiguration:^(__unused BTConfiguration * _Nullable configuration, __unused NSError * _Nullable error) {
        if (error != nil) {
            completionBlock(nil, nil, error);
            return;
        }

        NSDictionary *postParameters = [self metaParametersWithParameters:parameters forHTTPType:httpType];
        [[self httpForType:httpType] POST:endpoint parameters:postParameters completion:completionBlock];
    }];
    ```
    On completion, the HTTP client is chosen depending on the `httpType`. In that scenario, a `BTGraphQLHTTP` object is returned. This object is initialized during the `GET` mentioned above, using an URL provided in the merchant configuration that was fetched, `BTAPIClient` [`l.263`](https://github.com/braintree/braintree_ios/blob/master/BraintreeCore/BTAPIClient.m#L263):
    ```objective-c
    NSURL *graphQLBaseURL = [configuration.json[@"graphQL"][@"url"] asURL];
    if (!self.graphQL && graphQLBaseURL) {
        if (self.clientToken) {
            self.graphQL = [[BTGraphQLHTTP alloc] initWithBaseURL:graphQLBaseURL authorizationFingerprint:self.clientToken.authorizationFingerprint];
        } else if (self.tokenizationKey) {
            self.graphQL = [[BTGraphQLHTTP alloc] initWithBaseURL:graphQLBaseURL tokenizationKey:self.tokenizationKey];
        }
    }
  ```
  However, the json I get when fetching the merchant configuration doesn't contain any `graphQL` node:

  <img width="1000" src="https://user-images.githubusercontent.com/5849587/42758157-715800c8-88fa-11e8-9acf-23094a6fbefb.png">

  The `BTGraphQLHTTP` object is then `nil` when I try to send a message to make the `POST` request:
  ```objective-c
  [[self httpForType:httpType] POST:endpoint parameters:postParameters completion:completionBlock];
  ```
<br>

In conclusion, I am not sure what the source of the issue is. Am I missing some configuration in my Braintree account?

**Comment by demerino** [comment]
Thanks for the report - we're taking a look at this. We'll need to enable the graphql endpoint for you. Can you share your merchant ID so we can get it enabled?

**Comment by demerino** [comment]
@Nonepse  The fix for this is coming. Apologies for the delay and thanks for taking the time for the detailed writeup.

**Comment by Nonepse** [comment]
Thanks @demerino!
Do you still need my merchant ID?

**Comment by demerino** [comment]
Nope. The fix will be for all merchants.

**Comment by demerino** [comment]
@Nonepse This should be fixed now in 6.4.0 which requires the most recently release of `Braintree`. Let me know if that fixes the issue.

**Comment by demerino** [comment]
@Nonepse Can you confirm that the issue is resolved? If so, I'll close this issue. The web implementation is under development and Android will follow.

Thanks for your patience, everyone.

**Comment by Nonepse** [comment]
I will try to find a slot today to test it.

**Comment by Nonepse** [comment]
Hey @demerino, looks like it's working well 👍

However I think there is a small UX issue when the user deletes the last card in his Vault. The `BTVaultManagementViewController` screen is empty.
- I guess it's better to display a message indicating that there are no saved cards left.
- Or even dismiss the screen, since we go back to the `BTPaymentSelectionViewController` sheet, which is empty, implying all the cards were removed.

<br>
<img src="https://user-images.githubusercontent.com/5849587/42934044-20d8ff30-8b3e-11e8-8bfb-d6f9f558991e.gif" width="300">

**Comment by demerino** [comment]
👍 Glad it works and good point. Thanks for the feedback. I'll work with our designers/translators to see which option works best for us.

Looking forward to seeing your app in action! Closing this issue for now.

**Comment by muthu-smove** [comment]
Is there a callback / delegate / notification when the user deletes a specific card using the braintree UI. (`request.vaultManager = true`) ?

**Comment by demerino** [comment]
@muthu-smove Not at this time, is this a feature that you'd be interested in?

**Comment by muthu-smove** [comment]
@demerino Yes it would be really helpful to have a delegate to intimate of the deletions.

Presently `BTDropInController` has handler which is a callback when a card is added / tapped on.

It would be great if the `BTDropInController` has a delegate which has a set of methods, one to intimate additions, deletions, that way it would be more scalable when there are more actions in future. That can replace the existing handler callback. If that is a source breaking change, then it would be nice to handle delegate for deletion in addition to the existing handler.

**Comment by oxbits** [comment]
Is this available in Android yet?  Is there a ticket for it?

**Comment by bilalyaqoob** [comment]
hi, I want to let user to delete his saved payment cards I am using  this package
"react-native-braintree-payments-drop-in": "^1.2.0",

BraintreeDropIn.show({
      clientToken: this.state.clientToken,
      vaultManager: true <-- i also add this but it is not showing me edit option 
    }).then(result => { 

is there something i am missing

**Comment by sestevens** [comment]
@bilalyaqoob We don't maintain this React Native package, so we won't be much help with troubleshooting.

To allow customers to delete vaulted payment methods with the Braintree iOS Drop-in, you need to set this property to true: https://github.com/braintree/braintree-ios-drop-in/blob/bcb32b05021e9ed514a7d5fc81987565a0eb77db/BraintreeDropIn/Public/BTDropInRequest.h#L56

You'll also need to use a client token, not a tokenization key.
